File: Sources/Core/YAMLParser/AppParser.swift
import Foundation
import Yams

/// Parses YAML app definitions into typed schema models
public struct AppParser {

    public init() {}

    /// Parse a YAML string into an AppDefinition
    public func parse(_ yaml: String) throws -> AppDefinition {
        let decoder = YAMLDecoder()
        return try decoder.decode(AppDefinition.self, from: yaml)
    }

    /// Parse a YAML file at the given path
    public func parseFile(at path: String) throws -> AppDefinition {
        let url = URL(fileURLWithPath: path)
        let yaml = try String(contentsOf: url, encoding: .utf8)
        return try parse(yaml)
    }

    /// Parse a directory containing app.yaml and screen files in any subdirectory.
    ///
    /// Structure:
    /// ```
    /// myapp/
    ///   app.yaml              — app config, components, and optionally inline screens
    ///   screens/
    ///     home.yaml           — single ScreenDefinition per file
    ///     settings.yaml
    ///   screens/auth/
    ///     login.yaml          — nested subdirectories also scanned
    /// ```
    public func parseDirectory(at dirPath: String) throws -> AppDefinition {
        let fm = FileManager.default
        let appFile = (dirPath as NSString).appendingPathComponent("app.yaml")

        guard fm.fileExists(atPath: appFile) else {
            throw ParserError.missingAppYaml(directory: dirPath)
        }

        var app = try parseFile(at: appFile)

        let decoder = YAMLDecoder()

        let componentFiles = Self.findComponentFiles(in: dirPath)
        for filePath in componentFiles {
            let yaml = try String(contentsOfFile: filePath, encoding: .utf8)
            let component = try decoder.decode(CustomComponentDefinition.self, from: yaml)
            guard let name = component.name else { continue }
            if app.components == nil { app.components = [:] }
            app.components?[name] = component
        }

        let screenFiles = Self.findYAMLFiles(in: dirPath, excluding: "app.yaml")
        if !screenFiles.isEmpty {
            for filePath in screenFiles {
                let yaml = try String(contentsOfFile: filePath, encoding: .utf8)
                let screen = try decoder.decode(ScreenDefinition.self, from: yaml)
                app.screens.append(screen)
            }
        }

        return app
    }

    /// Merges a directory back into a single YAML string for hot reload broadcast.
    /// Reads app.yaml, component files, and all screen YAML files from subdirectories.
    public func mergeDirectoryToYAML(at dirPath: String) throws -> String {
        let appFile = (dirPath as NSString).appendingPathComponent("app.yaml")
        var appYaml = try String(contentsOfFile: appFile, encoding: .utf8)

        let componentFiles = Self.findComponentFiles(in: dirPath)
        if !componentFiles.isEmpty {
            if !appYaml.contains("\ncomponents:") && !appYaml.contains("\ncomponents :") {
                appYaml += "\ncomponents:\n"
            }
            for filePath in componentFiles {
                let compYaml = try String(contentsOfFile: filePath, encoding: .utf8)
                let lines = compYaml.components(separatedBy: "\n")
                var name: String?
                var bodyLines: [String] = []
                for line in lines {
                    let trimmed = line.trimmingCharacters(in: .whitespaces)
                    if trimmed.hasPrefix("name:") && name == nil {
                        name = trimmed.replacingOccurrences(of: "name:", with: "").trimmingCharacters(in: .whitespaces)
                    } else {
                        bodyLines.append(line)
                    }
                }
                guard let componentName = name, !componentName.isEmpty else { continue }
                let indentedBody = bodyLines
                    .map { $0.isEmpty ? "" : "    \($0)" }
                    .joined(separator: "\n")
                appYaml += "  \(componentName):\n\(indentedBody)\n"
            }
        }

        let screenFiles = Self.findYAMLFiles(in: dirPath, excluding: "app.yaml")
        if !screenFiles.isEmpty {
            if !appYaml.contains("\nscreens:") && !appYaml.contains("\nscreens :") {
                appYaml += "\nscreens:\n"
            }

            for filePath in screenFiles {
                let screenYaml = try String(contentsOfFile: filePath, encoding: .utf8)
                appYaml += "  - " + screenYaml.replacingOccurrences(
                    of: "\n",
                    with: "\n    "
                ) + "\n"
            }
        }

        return appYaml
    }

    /// Directories that should never be scanned for YAML files (build artifacts, etc.)
    private static let ignoredDirectoryNames: Set<String> = ["build", "node_modules", ".build", "DerivedData", "assets"]

    /// Whether a relative path passes through an ignored directory.
    private static func isInsideIgnoredDirectory(_ relative: String) -> Bool {
        let components = (relative as NSString).pathComponents
        return components.contains(where: { ignoredDirectoryNames.contains($0) })
    }

    /// Recursively finds all .yaml/.yml files in subdirectories of `dirPath`,
    /// excluding app.yaml files, component files, and build artifacts.
    /// Returns sorted absolute paths.
    public static func findYAMLFiles(in dirPath: String, excluding rootFile: String) -> [String] {
        let fm = FileManager.default
        guard let enumerator = fm.enumerator(atPath: dirPath) else { return [] }

        var results: [String] = []
        while let relative = enumerator.nextObject() as? String {
            let filename = (relative as NSString).lastPathComponent
            if filename == rootFile { continue }
            if filename.hasPrefix(".") { continue }
            if isInsideIgnoredDirectory(relative) { continue }
            guard relative.hasSuffix(".yaml") || relative.hasSuffix(".yml") else { continue }
            guard !relative.hasSuffix(".component.yaml") else { continue }

            results.append((dirPath as NSString).appendingPathComponent(relative))
        }
        return results.sorted()
    }

    /// Recursively finds all *.component.yaml files in `dirPath`.
    /// Returns sorted absolute paths.
    public static func findComponentFiles(in dirPath: String) -> [String] {
        let fm = FileManager.default
        guard let enumerator = fm.enumerator(atPath: dirPath) else { return [] }

        var results: [String] = []
        while let relative = enumerator.nextObject() as? String {
            let filename = (relative as NSString).lastPathComponent
            if filename.hasPrefix(".") { continue }
            if isInsideIgnoredDirectory(relative) { continue }
            guard relative.hasSuffix(".component.yaml") else { continue }

            results.append((dirPath as NSString).appendingPathComponent(relative))
        }
        return results.sorted()
    }

    /// Detect whether a path is a directory (multi-file project) or a single file
    public static func isDirectory(_ path: String) -> Bool {
        var isDir: ObjCBool = false
        FileManager.default.fileExists(atPath: path, isDirectory: &isDir)
        return isDir.boolValue
    }
}

/// Errors produced during YAML app definition parsing.
public enum ParserError: Error, LocalizedError {
    case missingAppYaml(directory: String)

    public var errorDescription: String? {
        switch self {
        case .missingAppYaml(let dir):
            return "No app.yaml found in directory: \(dir)"
        }
    }
}

File: Android/melody-core/src/main/java/com/melody/core/parser/AppParser.kt
package com.melody.core.parser

import android.content.Context
import com.charleskorn.kaml.Yaml
import com.charleskorn.kaml.YamlConfiguration
import com.melody.core.schema.AppDefinition
import com.melody.core.schema.CustomComponentDefinition
import com.melody.core.schema.ScreenDefinition
import java.io.File

class ParserError(message: String) : Exception(message)

/**
 * Parses YAML app definitions into typed schema models.
 * Port of iOS AppParser.swift.
 */
class AppParser {
    private val yaml = Yaml(
        configuration = YamlConfiguration(
            strictMode = false
        )
    )

    /** Parse a YAML string into an AppDefinition */
    fun parse(yamlString: String): AppDefinition {
        return yaml.decodeFromString(AppDefinition.serializer(), yamlString)
    }

    /** Parse a YAML file at the given path */
    fun parseFile(path: String): AppDefinition {
        val content = File(path).readText()
        return parse(content)
    }

    /**
     * Parse a directory containing app.yaml and screen files in subdirectories.
     */
    fun parseDirectory(dirPath: String): AppDefinition {
        val dir = File(dirPath)
        val appFile = File(dir, "app.yaml")

        if (!appFile.exists()) {
            throw ParserError("No app.yaml found in directory: $dirPath")
        }

        val app = parseFile(appFile.absolutePath)

        val componentFiles = findComponentFiles(dir)
        for (filePath in componentFiles) {
            val compYaml = File(filePath).readText()
            val component = yaml.decodeFromString(CustomComponentDefinition.serializer(), compYaml)
            val name = component.name ?: continue
            if (app.components == null) app.components = mutableMapOf()
            app.components!![name] = component
        }

        val screenFiles = findYAMLFiles(dir, "app.yaml")
        for (filePath in screenFiles) {
            val screenYaml = File(filePath).readText()
            val screen = yaml.decodeFromString(ScreenDefinition.serializer(), screenYaml)
            app.screens.add(screen)
        }

        return app
    }

    /**
     * Parse app definition from Android assets directory.
     * Loads app.yaml and all screen YAML files from subdirectories.
     */
    fun parseFromAssets(context: Context, rootPath: String = ""): AppDefinition {
        val assetManager = context.assets

        val appYamlPath = if (rootPath.isEmpty()) "app.yaml" else "$rootPath/app.yaml"
        val appYaml = assetManager.open(appYamlPath).bufferedReader().readText()
        val app = parse(appYaml)

        val componentsDir = if (rootPath.isEmpty()) "components" else "$rootPath/components"
        try {
            val componentAssetFiles = listAssetFiles(context, componentsDir)
            for (filePath in componentAssetFiles) {
                if (filePath.endsWith(".component.yaml")) {
                    try {
                        val compYaml = assetManager.open(filePath).bufferedReader().readText()
                        val component = yaml.decodeFromString(CustomComponentDefinition.serializer(), compYaml)
                        val name = component.name ?: continue
                        if (app.components == null) app.components = mutableMapOf()
                        app.components!![name] = component
                    } catch (e: Exception) {
                        android.util.Log.e("Melody", "Failed to parse component: $filePath", e)
                    }
                }
            }
        } catch (_: Exception) { }

        val screensDir = if (rootPath.isEmpty()) "screens" else "$rootPath/screens"
        try {
            val screenFiles = listAssetFiles(context, screensDir)
            for (filePath in screenFiles) {
                if ((filePath.endsWith(".yaml") || filePath.endsWith(".yml")) &&
                    !filePath.endsWith(".component.yaml")) {
                    try {
                        val screenYaml = assetManager.open(filePath).bufferedReader().readText()
                        val screen = yaml.decodeFromString(ScreenDefinition.serializer(), screenYaml)
                        app.screens.add(screen)
                    } catch (e: Exception) {
                        android.util.Log.e("Melody", "Failed to parse screen: $filePath", e)
                    }
                }
            }
        } catch (_: Exception) { }

        return app
    }

    /**
     * Merge directory back into a single YAML string (for hot reload).
     */
    fun mergeDirectoryToYAML(dirPath: String): String {
        val dir = File(dirPath)
        val appFile = File(dir, "app.yaml")
        var appYaml = appFile.readText()

        val componentFiles = findComponentFiles(dir)
        if (componentFiles.isNotEmpty()) {
            if (!appYaml.contains("\ncomponents:") && !appYaml.contains("\ncomponents :")) {
                appYaml += "\ncomponents:\n"
            }
            for (filePath in componentFiles) {
                val compYaml = File(filePath).readText()
                val lines = compYaml.split("\n")
                var name: String? = null
                val bodyLines = mutableListOf<String>()
                for (line in lines) {
                    val trimmed = line.trim()
                    if (trimmed.startsWith("name:") && name == null) {
                        name = trimmed.removePrefix("name:").trim()
                    } else {
                        bodyLines.add(line)
                    }
                }
                if (name.isNullOrEmpty()) continue
                val indentedBody = bodyLines.joinToString("\n") {
                    if (it.isEmpty()) "" else "    $it"
                }
                appYaml += "  $name:\n$indentedBody\n"
            }
        }

        val screenFiles = findYAMLFiles(dir, "app.yaml")
        if (screenFiles.isNotEmpty()) {
            if (!appYaml.contains("\nscreens:") && !appYaml.contains("\nscreens :")) {
                appYaml += "\nscreens:\n"
            }

            for (filePath in screenFiles) {
                val screenYaml = File(filePath).readText()
                appYaml += "  - " + screenYaml.replace("\n", "\n    ") + "\n"
            }
        }

        return appYaml
    }

    companion object {
        /**
         * Recursively find all YAML files in subdirectories, excluding a root file
         * and component files (*.component.yaml).
         */
        fun findYAMLFiles(dir: File, excludeRootFile: String): List<String> {
            val results = mutableListOf<String>()
            dir.walk()
                .filter { it.isFile }
                .filter { it.name != excludeRootFile }
                .filter { !it.name.startsWith(".") }
                .filter { it.extension == "yaml" || it.extension == "yml" }
                .filter { !it.name.endsWith(".component.yaml") }
                .forEach { results.add(it.absolutePath) }
            return results.sorted()
        }

        /**
         * Recursively find all *.component.yaml files in a directory.
         */
        fun findComponentFiles(dir: File): List<String> {
            val results = mutableListOf<String>()
            dir.walk()
                .filter { it.isFile }
                .filter { !it.name.startsWith(".") }
                .filter { it.name.endsWith(".component.yaml") }
                .forEach { results.add(it.absolutePath) }
            return results.sorted()
        }

        /**
         * Recursively list asset files under a given path.
         */
        private fun listAssetFiles(context: Context, path: String): List<String> {
            val results = mutableListOf<String>()
            val assetManager = context.assets
            val files = assetManager.list(path) ?: return results
            for (file in files) {
                val fullPath = "$path/$file"
                val subFiles = assetManager.list(fullPath)
                if (subFiles != null && subFiles.isNotEmpty()) {
                    results.addAll(listAssetFiles(context, fullPath))
                } else {
                    results.add(fullPath)
                }
            }
            return results.sorted()
        }
    }
}
